### ISSUE TITLE
Discussion: Concurrent IO in Core

### ISSUE THREAD

**Issue by omad** [opendatacube/datacube-core#625]
There's some excellent discussion in #624 about concurrent IO in datacube that needs to be made more visible. Lets continue in this thread, and use #467 for `save()` conversations.

@petewa has [described the direction CSIRO is thinking about based on their work using S3](https://github.com/opendatacube/datacube-core/issues/624#issuecomment-454610053)

@Kirill888 has been refactoring and cleaning `develop` to enable better concurrent IO support, but heading in a very different direction.

**Comment by petewa** [comment]
Pushing the concurrency down into the drivers was one out of 3 options we considered. It is possible but currently the drivers do not have enough information to make it work well and fuse_measurements will need to be decoupled out as a extra post_process step after the IO. This would make a lot of sense if we have nd-block reads supported by the index and drivers.

The following text is copied from #624, so may be a little out of context

The concurrency issue I'm trying to solve right now is, threadpool/processpool concurrency on creation, which is influenced by the cache. The way I did it was multi-tier, thread on time in core and space in the s3 driver. It was done like this to maintain compatibility with core, I somewhat regret this decision in hindsight. I should have done it properly even though it would cause more pain for GA during the s3 driver PR.

I'm thinking of moving the threadpool/process pool up further into core, possibly using/extending the executor class with pathos, but to do this properly there needs to be some sort of IO planning to prevent deadlocks.

For example:
for a dc.load():

    dc.metadata_for_load() - get all metadata for a dc.load
    dc.load_data()
        dc.create_storage - pre-create one contiguous array to hold all data.
        Plan all IO and produce a list of work items i.e. ([t, x, y, band, etc IO read] -> array slice)
        Send this to the core threadpool/processpool
        All work items happen in parallel and write/streams back into the pre-created array in parallel.
        pre-created array may be a shared memory array that may be in memory or disk and is accessible across threads & processed.

At least that's the direction I'm heading towards. I try to think of every function in an array operator sort of way so everything comes down to an array operator of sorts.

Whether I get time to work on it is a different story.

The above is similar to what happens in our Job Result Object (JRO) in our execution-engine branch for future computed results. It creates an array mapping that maps array slices to a list actual IO calls that can stream in parallel back to the result array.

I'll stop there, too much information to digest already.


-----------------------------------

My thoughts can be hard to follow :) I have trouble myself.

My thoughts is to make dc.save() happen in parallel with a simple array like call.
Either dc.save a whole xarray dataset/dataarray or a subset of an array (array slice).

For reads, s3aio (what the s3-driver uses) maps an array slice to a list of read IO calls to s3.
For writes, we can reuse that mapping to write back to s3 in parallel, because the small 5-10mb s3aio chunks/blocks are stored in different s3 objects, writes should be amazingly parallel and fast.
s3aio was designed for fast reads & writes (not implemented yet, function stub is there) given s3's limitation of a single change to a s3 object, re-writes the entire object. This is why we went with a chunk per s3 object, trading off some IO $ cost for more potential write performance.

s3aio could also be standalone library that maps a big array, sharded to s3 with parallel reads and writes via standard array slice operators.

Massively parallel writes back to s3, is what may set us a part, if done right.

If done right, it can do a parallel save across nodes as well and could also be used for easy ingest.

For example:
my big virtual array x[100000000000, 100000000000, 1000000000000, 1000000000000]
x is mapped across s3.
I want to save/ingest that into another form
I can do a dc.save() on x and it could operate on sub-slices per node
The final array shares the same mapping (very convenient) and can be indexed as usual with core like it happens now with the s3-driver.

Within a single node, it'll use the same threading model in the post further up.

Would anyone like to see this happen?

**Comment by petewa** [comment]
Concurrency issues aside, which option are you leaning towards? or maybe an option 4?

Option 1: Time IO in core, 2D Spatial IO in drivers.
* Bunch of 2D reads (2.5D approach) in drivers and "stacked" on time in core.
* The way it is now.

Option 2: Time + 2D spatial IO in drivers.
* Bunch of 2D reads (2.5D approach) and "stacked" on time in drivers.

* Drivers will need to know about time to do this well.
* The driver cache needs some work to make this play nice.

Option 3: nd IO in drivers.
* A single nD read (nD approach) in drivers that decomposes into many parallel read IOs.
  * Possible with:
     * `dc.metadata_for_load` and/or #319 which is similar.
     * storage units where time dimension size >= 1

Ideally we want more smarts in the drivers, currently time is not passed into the drivers in a meaningful way to combine reads if the storage units are truly multi-dimensional.

e.g. if my storage unit is (t=2, x=10, y=10) and my dc.load wants to load everything but it only knows about single time slices, so it does 2 reads (t=1, x=10, y=10). If the drivers is aware of the time component, it can do a single (t=2, x=10, y=10) read. A simple example.

fuse_measurements(fuse + re-project) should to be handled separately in all options.

**Comment by Kirill888** [comment]
Our plan currently is to move concurrency into the driver by changing driver/core API to be futures-based.

Driver will be responsible for managing IO threads. That way any number of concurrency back ends can be implemented by different drivers. So if your driver only needs one thread because it's using asyncIO for example, that's fine. If your driver demands multi-processing (netcdf library is not concurrent for example) than that is also possible, just harder to implement.

So far we still want to keep 1-time-slice reads, but you can have multiple concurrent reads since `read` will just return a promise of future data not actual data. Driver then has an option of coalescing time stacked reads if underlying data structures can benefit from it. Same goes for grouping reads across bands, driver choice: if complexity of figuring out what bands are stored together is worth the speedup you get by sharing single file open instead of just doing multiple (possibly concurrent) independent opens then do it, if not simple is good and with enough concurrency might be better anyway.

One important change we already made to driver interface is removal of `reproject`. Driver is now only responsible for reading data in the native projection, but we might request scaled imagery (needed to enable use of overviews). This makes some of your suggestions @petewa not possible in the generic case, I'm talking about pre-allocated shared memory, driver won't necessary write into final destination RAM.

Nothing stops driver implementor from using cross-process shared memory with multi-processing, but making that part of core/driver API contract would be a bad decision I think, mainly because you need to allocate shared memory before forking, and fork is costly. Also cross-platform issues of course.

**Comment by petewa** [comment]
Hi @Kirill888, thanks for sharing your thoughts.

Could you please go into more detail? It's hard to see the actual implications. On one hand it looks like it can work and on the other hand it may not work.

**Comment by Kirill888** [comment]
Change `open` and `read` to return futures instead of be blocking calls:

- `Datasource.open() -> Future[GeoRasterReader]`
- `GeoRasterReader.read(window, out_shape) -> Future[np.ndarray]`

Where:

https://github.com/opendatacube/datacube-core/blob/bbd41f74a94741d8922193d56d5de88505476759/datacube/drivers/datasource.py#L18

- We might get rid of `DataSource` and ask for `Future[GeoRasterReader]` directly from the driver
  - change `driver.DataSource(BandInfo).open()`
  -  to `driver.open(BandInfo)` 
- We might add some `Context` class so that driver can have a common place to store cache opened file handles for example:
  - `driver.create_context() -> Context` -- once per load
  - `driver.open(band_info, shared_ctx)`
- We might change `read` to accept destination `ndarray` instead of `out_shape` to avoid pixel copying.
- We will need some sort of configuration pass-through and global state management for individual drivers, so that driver can start threads/processes to be shared across loads, but not on start up and with possibility to clean up/reconfigure/restart.

So instead of 

```python
DataSource(band_info).open().read()
```

You'll need 

```python
DataSource(band_info).open().result().read().result()
```

to read a single band.

**Comment by Kirill888** [comment]
@petewa @rtaib @woodcockr  in preparation for driver change I have isolated S3AIO-based data loading into a separate sub-module in this PR #629.

This will allow us to have working S3AIO driver even as we change the driver interface to futures-based `open/read`. Also as part of the change `use_threads` parameter is re-instated, it will fall through from `.load|.load_data` to the `legacy.load_data`. I have not put back actual multi-threading, but I think it can be done quite easily by one of you guys.

https://github.com/opendatacube/datacube-core/blob/9ca148b1b074403f48a2908dd38f0de70efc4334/datacube/api/_legacy/load.py#L189-L196

My understanding is that you don't really need nor use multi-processing as multi-threading works just fine for S3 access, so code above can be made to run concurrently using `ThreadPool` without any further changes in `api.core`, you can parallelise across both time and measurement dimensions using one worker pool. No need to use shared memory array, no need to have nested thread pools, like you do in your fork. Can even lazily create and re-use thread pool for further improved throughput.

I think with this change implemented you can start using `develop` as your base again.

If you do need `SharedArray` for other reasons, this can still be achieved without changes in `api.core`, just supply `data_func` that allocates (but doesn't fill) `numpy` array backed by `SharedArray` storage to this function:

https://github.com/opendatacube/datacube-core/blob/9ca148b1b074403f48a2908dd38f0de70efc4334/datacube/api/core.py#L365
https://github.com/opendatacube/datacube-core/blob/9ca148b1b074403f48a2908dd38f0de70efc4334/datacube/api/core.py#L381-L384

**Comment by petewa** [comment]
Thanks @Kirill888, I'll have a look through the changes.

**Comment by Kirill888** [comment]
It was a bit too large for a comment so see below, document I prepared for todays discussion@petewa @omad @woodcockr @GypsyBojangles @rtaib 

https://gist.github.com/Kirill888/be5e61317ae2ea69935a0811fe971f89
